Incorrect thread safety in BatchSignatureVerifier

### Description
`BatchSignatureVerifier` stores jobs to be completed in an `ArrayList` without synchronisation but it's `verify` method is called from a parallel stream in `BlockProcessorUtil.verify_attestations`.  

I've seen two different exceptions result from this.

During a sync:
```
2020-05-17 23:24:25.665+10:00 | ForkJoinPool.commonPool-worker-31 | ERROR | BlockImporter | Internal error while importing block: BeaconBlock{root=0x82324321a68cf0fbd9600e84598081f80b622461d02b763801b5131d88233804, slot=139813, proposer_index=317, parent_root=0x733e820f22b8707a3235a000aee0baa21ce7b6c5b1063364b9795440eeab6fb6, state_root=0x519bbed9a085a08dbbbf0301e123c5d9d82b534908730cd6143eab7383ed15f2, body=0xd7e87833015d310f1250317c0699c2116e3d514cf55ad4468027f0e7639afe50}
java.lang.NullPointerException: null
	at tech.pegasys.teku.core.blockvalidator.BatchSignatureVerifier.lambda$batchVerify$0(BatchSignatureVerifier.java:71) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952) ~[?:?]
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926) ~[?:?]
	at java.util.stream.AbstractTask.compute(AbstractTask.java:327) ~[?:?]
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) ~[?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.helpCC(ForkJoinPool.java:1115) ~[?:?]
	at java.util.concurrent.ForkJoinPool.awaitJoin(ForkJoinPool.java:1687) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:411) ~[?:?]
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at tech.pegasys.teku.core.blockvalidator.BatchSignatureVerifier.batchVerify(BatchSignatureVerifier.java:72) ~[classes/:?]
	at tech.pegasys.teku.core.blockvalidator.BatchBlockValidator.validatePreState(BatchBlockValidator.java:40) ~[classes/:?]
	at tech.pegasys.teku.core.blockvalidator.BlockValidator.validate(BlockValidator.java:88) ~[classes/:?]
	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:100) ~[classes/:?]
	at tech.pegasys.teku.core.ForkChoiceUtil.on_block(ForkChoiceUtil.java:171) ~[classes/:?]
	at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.onBlock(ForkChoice.java:62) ~[classes/:?]
	at tech.pegasys.teku.statetransition.blockimport.BlockImporter.importBlock(BlockImporter.java:56) ~[classes/:?]
	at tech.pegasys.teku.sync.PeerSync.blockResponseListener(PeerSync.java:181) ~[classes/:?]
	at tech.pegasys.teku.networking.eth2.rpc.core.ResponseStreamImpl.respond(ResponseStreamImpl.java:66) ~[classes/:?]
	at tech.pegasys.teku.networking.eth2.rpc.core.AsyncResponseProcessor.lambda$processNextResponse$0(AsyncResponseProcessor.java:108) ~[classes/:?]
	at tech.pegasys.teku.util.async.SafeFuture.fromRunnable(SafeFuture.java:106) ~[classes/:?]
	at tech.pegasys.teku.util.async.AsyncRunner.lambda$runAsync$0(AsyncRunner.java:24) ~[classes/:?]
	at tech.pegasys.teku.util.async.DelayedExecutorAsyncRunner.lambda$runAsync$1(DelayedExecutorAsyncRunner.java:57) ~[classes/:?]
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426) [?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) [?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020) [?:?]
	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656) [?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594) [?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177) [?:?]
```

Regenerating hot states on startup:

```
Teku failed to start.
teku_1           | java.util.concurrent.CompletionException: java.lang.ArrayIndexOutOfBoundsException: Index 23 out of bounds for length 22
teku_1           | 	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
teku_1           | 	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.BatchBlockValidator.validatePreState(BatchBlockValidator.java:36)
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.BlockValidator.validate(BlockValidator.java:88)
teku_1           | 	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:100)
teku_1           | 	at tech.pegasys.teku.core.StateTransition.initiate(StateTransition.java:72)
teku_1           | 	at tech.pegasys.teku.storage.server.rocksdb.dataaccess.V3RocksDbDao.processBlock(V3RocksDbDao.java:233)
teku_1           | 	at tech.pegasys.teku.storage.server.rocksdb.dataaccess.V3RocksDbDao.initializeHotStates(V3RocksDbDao.java:206)
teku_1           | 	at tech.pegasys.teku.storage.server.rocksdb.dataaccess.V3RocksDbDao.initialize(V3RocksDbDao.java:176)
teku_1           | 	at tech.pegasys.teku.storage.server.rocksdb.dataaccess.V3RocksDbDao.<init>(V3RocksDbDao.java:60)
teku_1           | 	at tech.pegasys.teku.storage.server.rocksdb.RocksDbDatabase.createV3(RocksDbDatabase.java:61)
teku_1           | 	at tech.pegasys.teku.storage.server.rocksdb.RocksDbDatabase.createV3(RocksDbDatabase.java:57)
teku_1           | 	at tech.pegasys.teku.storage.server.VersionedDatabaseFactory.createV3Database(VersionedDatabaseFactory.java:70)
teku_1           | 	at tech.pegasys.teku.storage.server.VersionedDatabaseFactory.createDatabase(VersionedDatabaseFactory.java:58)
teku_1           | 	at tech.pegasys.teku.storage.server.ChainStorageServer.start(ChainStorageServer.java:53)
teku_1           | 	at tech.pegasys.teku.util.async.SafeFuture.fromRunnable(SafeFuture.java:106)
teku_1           | 	at tech.pegasys.teku.services.chainstorage.ChainStorageService.doStart(ChainStorageService.java:39)
teku_1           | 	at tech.pegasys.teku.service.serviceutils.Service.start(Service.java:33)
teku_1           | 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
teku_1           | 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
teku_1           | 	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:517)
teku_1           | 	at tech.pegasys.teku.services.ServiceController.doStart(ServiceController.java:44)
teku_1           | 	at tech.pegasys.teku.service.serviceutils.Service.start(Service.java:33)
teku_1           | 	at tech.pegasys.teku.BeaconNode.start(BeaconNode.java:92)
teku_1           | 	at tech.pegasys.teku.Teku.start(Teku.java:38)
teku_1           | 	at tech.pegasys.teku.cli.BeaconNodeCommand.call(BeaconNodeCommand.java:241)
teku_1           | 	at tech.pegasys.teku.cli.BeaconNodeCommand.call(BeaconNodeCommand.java:61)
teku_1           | 	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
teku_1           | 	at picocli.CommandLine.access$900(CommandLine.java:145)
teku_1           | 	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
teku_1           | 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
teku_1           | 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
teku_1           | 	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1975)
teku_1           | 	at picocli.CommandLine.execute(CommandLine.java:1904)
teku_1           | 	at tech.pegasys.teku.cli.BeaconNodeCommand.parse(BeaconNodeCommand.java:194)
teku_1           | 	at tech.pegasys.teku.Teku.main(Teku.java:30)
teku_1           | Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 23 out of bounds for length 22
teku_1           | 	at java.base/java.util.ArrayList.add(ArrayList.java:486)
teku_1           | 	at java.base/java.util.ArrayList.add(ArrayList.java:498)
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.BatchSignatureVerifier.verify(BatchSignatureVerifier.java:55)
teku_1           | 	at tech.pegasys.teku.datastructures.util.AttestationUtil.is_valid_indexed_attestation(AttestationUtil.java:170)
teku_1           | 	at tech.pegasys.teku.core.BlockProcessorUtil.lambda$verify_attestations$2(BlockProcessorUtil.java:469)
teku_1           | 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
teku_1           | 	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1631)
teku_1           | 	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
teku_1           | 	at java.base/java.util.stream.FindOps$FindTask.doLeaf(FindOps.java:319)
teku_1           | 	at java.base/java.util.stream.AbstractShortCircuitTask.compute(AbstractShortCircuitTask.java:115)
teku_1           | 	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.helpCC(ForkJoinPool.java:1115)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinPool.externalHelpComplete(ForkJoinPool.java:1957)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinTask.tryExternalHelp(ForkJoinTask.java:378)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinTask.externalAwaitDone(ForkJoinTask.java:323)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:412)
teku_1           | 	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
teku_1           | 	at java.base/java.util.stream.FindOps$FindOp.evaluateParallel(FindOps.java:160)
teku_1           | 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
teku_1           | 	at java.base/java.util.stream.ReferencePipeline.findAny(ReferencePipeline.java:548)
teku_1           | 	at tech.pegasys.teku.core.BlockProcessorUtil.verify_attestations(BlockProcessorUtil.java:471)
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.SimpleBlockValidator.validatePreState(SimpleBlockValidator.java:78)
teku_1           | 	at tech.pegasys.teku.core.blockvalidator.BatchBlockValidator.validatePreState(BatchBlockValidator.java:33)
teku_1           | 	... 37 more
```

--- a/ethereum/core/src/main/java/tech/pegasys/teku/core/blockvalidator/BatchSignatureVerifier.java
+++ b/ethereum/core/src/main/java/tech/pegasys/teku/core/blockvalidator/BatchSignatureVerifier.java
@@ -13,6 +13,7 @@
 
 package tech.pegasys.teku.core.blockvalidator;
 
+import com.google.common.annotations.VisibleForTesting;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
@@ -30,6 +31,8 @@ import tech.pegasys.teku.bls.mikuli.BLS12381.BatchSemiAggregate;
  *
  * <p>Every instance of this class is disposable, i.e. it is intended for just a single batch and a
  * single {@link #batchVerify()} call.
+ *
+ * <p>This is thread-safe class.
  */
 public class BatchSignatureVerifier implements BLSSignatureVerifier {
   private static class Job {
@@ -46,11 +49,12 @@ public class BatchSignatureVerifier implements BLSSignatureVerifier {
     }
   }
 
-  private final List<Job> toVerify = new ArrayList<>();
+  @VisibleForTesting final List<Job> toVerify = new ArrayList<>();
   private boolean complete = false;
 
   @Override
-  public boolean verify(List<BLSPublicKey> publicKeys, Bytes message, BLSSignature signature) {
+  public synchronized boolean verify(
+      List<BLSPublicKey> publicKeys, Bytes message, BLSSignature signature) {
     if (complete) throw new IllegalStateException("Reuse of disposable instance");
     toVerify.add(new Job(toVerify.size(), publicKeys, message, signature));
     return true;
@@ -63,7 +67,7 @@ public class BatchSignatureVerifier implements BLSSignatureVerifier {
    * <p>After this method completes the instance should be disposed and any subsequent calls to this
    * instance methods would fail with exception
    */
-  public boolean batchVerify() {
+  public synchronized boolean batchVerify() {
     if (complete) throw new IllegalStateException("Reuse of disposable instance");
     List<BatchSemiAggregate> batchSemiAggregates =
         toVerify.stream()

--- a/ethereum/core/src/test/java/tech/pegasys/teku/core/blockvalidator/BatchSignatureVerifierTest.java
+++ b/ethereum/core/src/test/java/tech/pegasys/teku/core/blockvalidator/BatchSignatureVerifierTest.java
@@ -1,0 +1,70 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.core.blockvalidator;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.tuweni.bytes.Bytes;
+import org.junit.jupiter.api.Test;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.bls.BLSSignature;
+
+public class BatchSignatureVerifierTest {
+
+  @Test
+  public void testParallel() throws Exception {
+    BatchSignatureVerifier verifier = new BatchSignatureVerifier();
+
+    BLSPublicKey publicKey = BLSPublicKey.random(42);
+    Bytes message = Bytes.wrap("Hello, world!".getBytes(UTF_8));
+    BLSSignature signature = BLSSignature.random(42);
+
+    int nThreads = 64;
+    int iterations = 64;
+    ExecutorService executor = Executors.newFixedThreadPool(nThreads);
+    CountDownLatch latch = new CountDownLatch(nThreads);
+    List<Future<Boolean>> futures =
+        IntStream.range(0, nThreads)
+            .mapToObj(
+                __ ->
+                    executor.submit(
+                        () -> {
+                          latch.countDown();
+                          latch.await();
+                          boolean ret = true;
+                          for (int i = 0; i < iterations; i++) {
+                            ret &=
+                                verifier.verify(
+                                    Collections.singletonList(publicKey), message, signature);
+                          }
+                          return ret;
+                        }))
+            .collect(Collectors.toList());
+
+    for (Future<Boolean> future : futures) {
+      assertThat(future.get()).isTrue();
+    }
+    assertThat(verifier.toVerify.size()).isEqualTo(nThreads * iterations);
+    assertThat(verifier.toVerify).doesNotContainNull();
+  }
+}

